tests/e2e/postgres/locking_test.go

```diff
@@ -0,0 +1,227 @@
+package postgres_test
+
+import (
+	"context"
+	"errors"
+	"sort"
+	"sync"
+	"testing"
+
+	"github.com/pressly/goose/v4"
+	"github.com/pressly/goose/v4/internal/check"
+	"golang.org/x/sync/errgroup"
+)
+
+func TestLockModeAdvisorySession(t *testing.T) {
+	t.Parallel()
+
+	// The migrations are written in such a way that they cannot be applied concurrently. This test
+	// ensures that the advisory session lock mode works as expected.
+
+	options := goose.DefaultOptions().
+		SetDir(migrationsDir).
+		SetVerbose(testing.Verbose()).
+		SetLockMode(goose.LockModeAdvisorySession) /* ---------------- advisory session lock mode */
```

Comment:
If this is not set, these tests are guaranteed to fail, and we add checkpoints between the 4 test cases.

I ran this test 100 times in a loop with and without the session-level advisory lock and it was:

```sh
# with locking
Success: 100 Fail: 0

# without locking
Success: 100 Fail: 0
```

Jotting down the errors we'd expect below. Now, we're not guaranteed to get all of the errors. 


```
unexpected error: failed to run SQL migration: 00001_a.sql: ERROR: duplicate key value violates unique constraint "pg_type_typname_nsp_index" (SQLSTATE 23505)

// or 

unexpected error: failed to run SQL migration: 00001_a.sql: ERROR: type "owner_type" already exists (SQLSTATE 42710)

// or 

unexpected error: failed to run SQL migration: 00010_j.sql: ERROR: deadlock detected (SQLSTATE 40P01)
```

That last one is interesting because migration `00010_j.sql` has `-- +goose NO TRANSACTION` set and attempts to run:

```sql
CREATE UNIQUE INDEX CONCURRENTLY ON owners(owner_name);
```

This will always raise a deadlock because of the way concurrent indexes are rebuilt in postrges. 


---

This was the script:

```sh
#!/bin/bash

SUCCESS=0
FAIL=0
for i in {1..100}
do
    echo "Running $i"
    go test -timeout 30s -run \
        ^TestLockModeAdvisorySession$ github.com/pressly/goose/v4/tests/e2e/postgres \
        -race -count=1 > /dev/null 2>&1
    case $? in
        0) 
            (( SUCCESS++ ))
            ;;
        1) 
            (( FAIL++ ))
            ;;
        *) 
            echo "Unknown"
            exit 1
            ;;
    esac
done

echo "Success: $SUCCESS Fail: $FAIL"
```